expo/eas-build
# [ENG-9010][ENG-8949][steps] Add support for custom TS/JS functions

# Why

https://linear.app/expo/issue/ENG-9010/allow-users-to-create-custom-tsjs-functions

# How

Allowed for the `path` property in the function config, which should point to a custom JS/TS module.

Added the mechanism for the `@expo/steps` library to use the user-defined JS/TS functions. I achieved it by doing `spawnAsync("node", [pathToSomeWrapperScript, pathToModuleWithCustomFunction, serializedFunctionArguments])` with wrapper script taking care of the deserialization of arguments, transforming them back into class instances, calling user function using these arguments and collecting and updating outputs and env variables at the end of function execution as suggested in https://github.com/expo/eas-custom-builds-example/pull/7.

Improved typing of `BuildStepInput` and `BuildStepOutput` classes, so the input value is inferred based on `required` and `allowedValueTypeName` types or `T` (for type) and `R` for required template arguments.

# Test Plan

## Manual tests

Used
```yml
build:
  name: Custom JS/TS functions
  steps:
    - my_js_fn:
        inputs:
          name: Dominik
          num: 3
          obj:
            foo: bar
    - my_ts_fn:
        id: my_ts_fn_id
        inputs:
          name: Wojtek
          num: 4
          obj:
            foo: baz
    - run:
        inputs:
          name: ${ steps.my_ts_fn_id.name }
          num: ${ steps.my_ts_fn_id.num }
          obj: ${ steps.my_ts_fn_id.obj }
        name: Verify if the outputs and env set
        command: |
          echo ${ inputs.name }
          echo ${ inputs.num }
          echo ${ inputs.obj }
          echo $MY_ENV_VAR


functions:
  my_js_fn:
    name: my-js-fn
    inputs:
      - name: name
      - name: num
        type: number
      - name: obj
        type: json
    path: ./test-custom-js-function

  my_ts_fn:
    name: my-ts-fn
    inputs:
      - name: name
      - name: num
        type: number
      - name: obj
        type: json
    outputs:
      - name: name
      - name: num
      - name: obj
    path: ./test-custom-ts-function
```
config for testing purposes, with two test JS/TS modules created with the following structure
<img width="354" alt="Screenshot 2023-07-26 at 14 37 58" src="https://github.com/expo/eas-build/assets/55145344/e9e497ed-3999-4dec-995d-e3f4fa61ea4a">

Custom TS function body (`index.ts`):
```ts
import {
  BuildStepContext,
  BuildStepInput,
  BuildStepInputValueTypeName,
  BuildStepOutput,
  BuildStepEnv,
} from '@expo/steps';

interface MyTsFunctionInputs {
  name: BuildStepInput<BuildStepInputValueTypeName.STRING, true>;
  num: BuildStepInput<BuildStepInputValueTypeName.NUMBER, true>;
  obj: BuildStepInput<BuildStepInputValueTypeName.JSON, true>;
}

interface MyTsFunctionOutputs {
  name: BuildStepOutput<true>;
  num: BuildStepOutput<true>;
  obj: BuildStepOutput<true>;
}

async function myTsFunction(
  ctx: BuildStepContext,
  {
    inputs,
    outputs,
    env,
  }: {
    inputs: MyTsFunctionInputs;
    outputs: MyTsFunctionOutputs;
    env: BuildStepEnv;
  }
) {
  ctx.logger.info('Running my custom TS function');
  ctx.logger.info(`Hello, ${inputs.name.value}!}`);
  ctx.logger.info(`Your number is ${inputs.num.value}`);
  ctx.logger.info(`Your object is ${JSON.stringify(inputs.obj.value)}`);
  ctx.logger.info('Done running my custom TS function');
  ctx.logger.warn('Warning from my custom TS function');
  ctx.logger.error('Error from my custom TS function');
  ctx.logger.info('Running a command');
  ctx.logger.debug('Debugging a command');
  ctx.logger.fatal('Fatal error from my custom TS function');

  ctx.logger.info('Setting outputs');
  outputs.name.set('Brent');
  outputs.num.set('123');
  outputs.obj.set(JSON.stringify({ foo: 'bar' })); // TODO: add support for other types of outputs different then string

  ctx.logger.info('Setting env vars');
  env['MY_ENV_VAR'] = 'my-value';
}

export default myTsFunction;
```

Custom JS function body (`index.js`):
```js
function myCustomJsFunction(ctx, { inputs }) {
  ctx.logger.info('Running my custom JS function');
  ctx.logger.info(`Hello, ${inputs.name.value}!}`);
  ctx.logger.info(`Your number is ${inputs.num.value}`);
  ctx.logger.info(`Your object is ${JSON.stringify(inputs.obj.value)}`);
  ctx.logger.info('Done running my custom JS function');
}

export default myCustomJsFunction;
```

<img width="802" alt="Screenshot 2023-07-26 at 14 43 35" src="https://github.com/expo/eas-build/assets/55145344/a2ba7b45-363f-4c79-98eb-ee000952f1ac">


## Automated tests

Add automated tests
